exporter/exporterhelper/internal/queue_capacity.go

```diff
@@ -0,0 +1,92 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package internal // import "go.opentelemetry.io/collector/exporter/exporterhelper/internal"
+
+import (
+	"sync/atomic"
+)
+
+type itemsCounter interface {
+	ItemsCount() int
+}
+
+// QueueCapacityLimiter is an interface to control the capacity of the queue.
+// This API is at the early stage of development and may change without backward compatibility
+// until https://github.com/open-telemetry/opentelemetry-collector/issues/8122 is resolved.
+type QueueCapacityLimiter[T any] interface {
+	// Claim tries to claim capacity for the given element. If the capacity is not available, it returns false.
+	Claim(T) bool
+	// Release releases capacity for the given queue element.
+	Release(T)
+	// Size is current size of the queue.
+	Size() int
+	// Capacity is the maximum capacity of the queue.
+	Capacity() int
+}
+
+type baseQueueCapacityLimiter struct {
+	size     *atomic.Uint64
+	capacity uint64
+}
+
+func newBaseQueueCapacityLimiter(capacity int) *baseQueueCapacityLimiter {
+	return &baseQueueCapacityLimiter{
+		size:     &atomic.Uint64{},
+		capacity: uint64(capacity),
+	}
+}
+
+func (bqcl *baseQueueCapacityLimiter) Size() int {
+	return int(bqcl.size.Load())
+}
+
+func (bqcl *baseQueueCapacityLimiter) Capacity() int {
+	return int(bqcl.capacity)
+}
+
+// requestCapacityLimiter is a capacity limiter that limits the queue based on the number of requests.
+type requestsCapacityLimiter[T any] struct {
+	baseQueueCapacityLimiter
+}
+
+func NewRequestsCapacityLimiter[T any](capacity int) QueueCapacityLimiter[T] {
+	return &requestsCapacityLimiter[T]{
+		baseQueueCapacityLimiter: *newBaseQueueCapacityLimiter(capacity),
+	}
+}
+
+func (rcl *requestsCapacityLimiter[T]) Claim(_ T) bool {
+	if rcl.size.Load() >= rcl.capacity {
+		return false
+	}
+	rcl.size.Add(1)
+	return true
+}
+
+func (rcl *requestsCapacityLimiter[T]) Release(_ T) {
+	rcl.size.Add(^uint64(0))
```

Comment:
This can overflow from 0 to 4294967295. 

I tested the new changes with persistent queue and saw that:
- when persistent queue initialize, it loads some `currentlyDispatchedItems` and consume it. Once an item is consumed, this number becomes overflow and all future `Claim` becomes blocked.

Would it make sense to safe guard it from overflowing(always above 0)? Otherwise it is hard to ensure callers to not accidentally call `Release`.

Reply:
Thanks for the reports! It's a bug indeed. Let me fix

Reply:
I implemented the more resilient queue size tracking for the persistent queue